### ISSUE TITLE
General: Qt cannot find its platform plugin anymore

### ISSUE THREAD

**Issue by dbrnz** [opencor/opencor#2089]
On Ubuntu 18, with the latest snapshot from opencor.ws:
1. Rename (or delete) the directory containing the Qt installation.
2. Reboot.
3. Attempt to start OpenCOR.
```
dave@ubuntu18:~/OpenCOR-latest$ ./OpenCOR 
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.

Aborted (core dumped)
```
Windows has the same issue; I've yet to test macOS.

**Comment by agarny** [comment]
?!?! Ok, I am going to have to check that too, I guess... ?!?!

P.S.: `Bug` and `Regression` is not suitable here. They should only be used if the issue you are reporting had been introduced in, say, OpenCOR 0.6. IOW, those labels are only for official versions of OpenCOR. This is to make things easier for me when compiling the what is new list of items.

**Comment by dbrnz** [comment]
Found it!
```
dave@ubuntu18:~/OpenCOR-latest$ ./OpenCOR 
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.

Aborted (core dumped)
dave@ubuntu18:~/OpenCOR-latest$ cd bin
dave@ubuntu18:~/OpenCOR-latest/bin$ ln -s ../plugins/platforms platforms
dave@ubuntu18:~/OpenCOR-latest/bin$ ls -l
total 5172
-rwxr-xr-x 1 dave dave 5294320 May 13 13:09 OpenCOR
lrwxrwxrwx 1 dave dave      20 May 18 12:23 platforms -> ../plugins/platforms
dave@ubuntu18:~/OpenCOR-latest/bin$ cd ..
dave@ubuntu18:~/OpenCOR-latest$ ./OpenCOR 
dave@ubuntu18:~/OpenCOR-latest$ rm bin/platforms
dave@ubuntu18:~/OpenCOR-latest$ ./OpenCOR 
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.

Aborted (core dumped)
dave@ubuntu18:~/OpenCOR-latest$ 
```
From the bottom of https://doc.qt.io/qt-5/linux-deployment.html:

> All Qt GUI applications require a plugin that implements the Qt Platform Abstraction (QPA) layer in Qt 5. For Linux/X11, the name of the platform plugin is libqxcb.so. This file must be located within a specific subdirectory (by default, platforms) **_under your distribution directory_**. Alternatively, it is possible to adjust the search path Qt uses to find its plugins, as described below.

Here `distribution directory` means the directory containing the executable.

Should we be using a [qt.conf](https://doc.qt.io/qt-5/qt-conf.html) file??

**Comment by dbrnz** [comment]
Another solution is to set `QT_QPA_PLATFORM_PLUGIN_PATH` to the `./plugins/platforms` directory.

It's not clear how to use `qt.conf` to just set the platform path. From my testing it appears that without QT_QPA_PLATFORM_PLUGIN_PATH set, `./platforms/` needs to be in both `./bin` and `./plugins`... Aargh.

**Comment by agarny** [comment]
I am surprised about this issue popping up now. I mean, I have known about Qt applications needing the Qt platform plugin since the beginning of OpenCOR. This is the reason we ship it with OpenCOR. So, I really don't understand why we are having this issue now...!?

What is certain is that `qt.conf` has been tried before and has caused us problems (see issue #376). So, definitely not an option, not to mention that I don't like having files around to "configure" things.

I can't say that I am happy about using `QT_QPA_PLATFORM_PLUGIN_PATH` either. I have seen your PR #2090, but what happens if you run OpenCOR by double clicking on the executable?

Bottom line: something somehow got broken recently while everything used to work fine. So, I want to understand the cause rather than just provide a fix that seems to solve the problem. So, more on this on Monday... :) (but happy for you to do whatever you think is right for OpenCOR+Python at this stage)

**Comment by dbrnz** [comment]
Totally agree with finding and understanding why things got broken.

Maybe the "worst case" solution is to programmatically set `QT_QPA_PLATFORM_PLUGIN_PATH` before Qt goes looking for its platform plugin, similar to calling `setLibraryPaths()`.

**Comment by agarny** [comment]
Yes, I would be ok with programatically setting `QT_QPA_PLATFORM_PLUGIN_PATH`, if setting that environment variable is the way to go. Definitely better than doing it in our bash/batch scripts.

**Comment by agarny** [comment]
Ok, if anything, I can indeed confirm the issue on both Windows and Linux (Ubuntu 18.04 LTS). OpenCOR 0.6 is fine (as expected since I checked everything back then) as are the `2019-02-01` and `2019-03-07` snapshots, but starting from the `2019-05-05` snapshot, I am getting the following error message on both Windows and Ubuntu:

```
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.
```

Between the `2019-03-07` and `2019-05-05` snapshots, we upgraded to Qt 5.12.3 LTS (from Qt 5.12.1 LTS). We also upgraded our minimum requirement for Ubuntu, from 16.04 LTS to 18.04 LTS. We also went from Mesa 18.x to Mesa 19.x. We also use MSVC 2017 now, on Windows.

The only thing that is common to both Windows and Linux is our Qt upgrade...!?

I have checked dependencies for `OpenCOR.exe` on Windows and can't see any difference. On Ubuntu, OpenCOR now needs a new system library (`libtinfo.so.5`).

I have also had a quick look at the changelog between Qt 5.12.1 and 5.12.3, but it's not easy to search and the quick searches that I did were unsuccessful in bringing up something relevant.

So, in the end, I have no idea about what is responsible for our problem...!? I am therefore going to rely on `QT_QPA_PLATFORM_PLUGIN_PATH`...

**Comment by agarny** [comment]
I thought that a simple call to `qputenv()` in our `initApplication()` method would do the trick and it does on Linux, but not on Windows.

This is because the call to `qputenv()` is too late (so, why does it work on Linux?!). In other words, if we call `qputenv` at the beginning of our `main()` procedure then OpenCOR fires up properly, but... the styling is not right. Indeed, OpenCOR looks as follows:

![New](https://user-images.githubusercontent.com/602265/57996445-9ecbca80-7b1b-11e9-81e4-0745e019f10d.jpg)

while it ought to look as follows:

![Old](https://user-images.githubusercontent.com/602265/57996463-b440f480-7b1b-11e9-9b19-5fdfe327a9b0.jpg)

**Comment by agarny** [comment]
Ok, I thought the "issue" was with upgrading from Qt 5.12.1 to 5.12.3, but I have just rebuilt OpenCOR using Qt 5.12.1 on Linux and... still the same message!?!?

```
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.
```

**Comment by agarny** [comment]
The same holds true on Windows, i.e. building OpenCOR (`master`) using Qt 5.12.1 results in the same message.

So, it looks like it might not be a Qt issue after all. However, something that I have noticed on both Windows and Linux is that building OpenCOR using Qt 5.12.1, we get the following build output:

```
-- The C compiler identification is GNU 7.4.0
-- The CXX compiler identification is GNU 7.4.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
Building a release version of OpenCOR using Qt 5.12.1 LTS...
/home/alan/OpencOR/src/checkforupdatesdialog.cpp:291: Class 'OpenCOR' lacks Q_OBJECT macro
/home/alan/OpencOR/src/checkforupdatesdialog.cpp:119: Unbalanced opening parenthesis in C++ code (or abuse of the C++ preprocessor)
/home/alan/OpencOR/src/misc/cliutils.cpp:177: Unbalanced opening parenthesis in C++ code (or abuse of the C++ preprocessor)
/home/alan/OpencOR/src/plugins/editing/CellMLAnnotationView/src/cellmlannotationviewmetadataeditdetailswidget.cpp:593: Excess closing parenthesis in C++ code (or abuse of the C++ preprocessor)
/home/alan/OpencOR/src/plugins/editing/CellMLAnnotationView/src/cellmlannotationviewmetadataeditdetailswidget.cpp:883: Class 'OpenCOR::CellMLAnnotationView' lacks Q_OBJECT macro
/home/alan/OpencOR/src/plugins/editing/CellMLAnnotationView/src/cellmlannotationviewmetadataeditdetailswidget.cpp:606: Unbalanced opening parenthesis in C++ code (or abuse of the C++ preprocessor)
/home/alan/OpencOR/src/plugins/editing/CellMLAnnotationView/src/cellmlannotationviewmetadatanormalviewdetailswidget.cpp:292: Excess closing parenthesis in C++ code (or abuse of the C++ preprocessor)
/home/alan/OpencOR/src/plugins/editing/CellMLAnnotationView/src/cellmlannotationviewmetadatanormalviewdetailswidget.cpp:305: Unbalanced opening parenthesis in C++ code (or abuse of the C++ preprocessor)
/home/alan/OpencOR/src/plugins/editing/CellMLAnnotationView/src/cellmlannotationviewplugin.cpp:321: Class 'OpenCOR::CellMLAnnotationView' lacks Q_OBJECT macro
/home/alan/OpencOR/src/plugins/editing/CellMLAnnotationView/src/cellmlannotationviewplugin.cpp:47: Unbalanced opening parenthesis in C++ code (or abuse of the C++ preprocessor)
/home/alan/OpencOR/src/plugins/editing/CellMLEditingView/src/cellmleditingviewplugin.cpp:237: Class 'OpenCOR::CellMLEditingView' lacks Q_OBJECT macro
/home/alan/OpencOR/src/plugins/editing/CellMLEditingView/src/cellmleditingviewplugin.cpp:48: Unbalanced opening parenthesis in C++ code (or abuse of the C++ preprocessor)
/home/alan/OpencOR/src/plugins/editing/CellMLTextView/src/cellmltextviewlexer.cpp:206: Excess closing parenthesis in C++ code (or abuse of the C++ preprocessor)
/home/alan/OpencOR/src/plugins/editing/CellMLTextView/src/cellmltextviewlexer.cpp:206: Unterminated C++ string
/home/alan/OpencOR/src/plugins/editing/CellMLTextView/src/cellmltextviewplugin.cpp:431: Class 'OpenCOR::CellMLTextView' lacks Q_OBJECT macro
/home/alan/OpencOR/src/plugins/editing/CellMLTextView/src/cellmltextviewplugin.cpp:47: Unbalanced opening parenthesis in C++ code (or abuse of the C++ preprocessor)
/home/alan/OpencOR/src/plugins/editing/CellMLTextView/src/cellmltextviewwidget.cpp:1243: Unbalanced opening parenthesis in C++ code (or abuse of the C++ preprocessor)
/home/alan/OpencOR/src/plugins/editing/RawCellMLView/src/rawcellmlviewplugin.cpp:381: Class 'OpenCOR::RawCellMLView' lacks Q_OBJECT macro
/home/alan/OpencOR/src/plugins/editing/RawCellMLView/src/rawcellmlviewplugin.cpp:47: Unbalanced opening parenthesis in C++ code (or abuse of the C++ preprocessor)
/home/alan/OpencOR/src/plugins/editing/RawCellMLView/src/rawcellmlviewwidget.cpp:388: Unbalanced opening parenthesis in C++ code (or abuse of the C++ preprocessor)
/home/alan/OpencOR/src/plugins/editing/RawSEDMLView/src/rawsedmlviewplugin.cpp:378: Class 'OpenCOR::RawSEDMLView' lacks Q_OBJECT macro
/home/alan/OpencOR/src/plugins/editing/RawSEDMLView/src/rawsedmlviewplugin.cpp:47: Unbalanced opening parenthesis in C++ code (or abuse of the C++ preprocessor)
/home/alan/OpencOR/src/plugins/editing/SEDMLEditingView/src/sedmleditingviewplugin.cpp:237: Class 'OpenCOR::SEDMLEditingView' lacks Q_OBJECT macro
/home/alan/OpencOR/src/plugins/editing/SEDMLEditingView/src/sedmleditingviewplugin.cpp:48: Unbalanced opening parenthesis in C++ code (or abuse of the C++ preprocessor)
/home/alan/OpencOR/src/plugins/miscellaneous/Core/src/corecliutils.cpp:907: Excess closing parenthesis in C++ code (or abuse of the C++ preprocessor)
/home/alan/OpencOR/src/plugins/miscellaneous/Core/src/corecliutils.cpp:913: Unbalanced opening parenthesis in C++ code (or abuse of the C++ preprocessor)
/home/alan/OpencOR/src/plugins/organisation/PMRWindow/src/pmrwindowplugin.cpp:63: Class 'OpenCOR::PMRWindow' lacks Q_OBJECT macro
/home/alan/OpencOR/src/plugins/organisation/PMRWindow/src/pmrwindowplugin.cpp:46: Unbalanced opening parenthesis in C++ code (or abuse of the C++ preprocessor)
/home/alan/OpencOR/src/plugins/organisation/PMRWorkspacesWindow/src/pmrworkspaceswindowplugin.cpp:67: Class 'OpenCOR::PMRWorkspacesWindow' lacks Q_OBJECT macro
/home/alan/OpencOR/src/plugins/organisation/PMRWorkspacesWindow/src/pmrworkspaceswindowplugin.cpp:50: Unbalanced opening parenthesis in C++ code (or abuse of the C++ preprocessor)
/home/alan/OpencOR/src/plugins/solver/CVODESolver/src/cvodesolverplugin.cpp:38: Unbalanced opening parenthesis in C++ code (or abuse of the C++ preprocessor)
/home/alan/OpencOR/src/plugins/solver/ForwardEulerSolver/src/forwardeulersolverplugin.cpp:38: Unbalanced opening parenthesis in C++ code (or abuse of the C++ preprocessor)
/home/alan/OpencOR/src/plugins/solver/FourthOrderRungeKuttaSolver/src/fourthorderrungekuttasolverplugin.cpp:38: Unbalanced opening parenthesis in C++ code (or abuse of the C++ preprocessor)
/home/alan/OpencOR/src/plugins/solver/HeunSolver/src/heunsolverplugin.cpp:38: Unbalanced opening parenthesis in C++ code (or abuse of the C++ preprocessor)
/home/alan/OpencOR/src/plugins/solver/KINSOLSolver/src/kinsolsolverplugin.cpp:38: Unbalanced opening parenthesis in C++ code (or abuse of the C++ preprocessor)
/home/alan/OpencOR/src/plugins/solver/SecondOrderRungeKuttaSolver/src/secondorderrungekuttasolverplugin.cpp:38: Unbalanced opening parenthesis in C++ code (or abuse of the C++ preprocessor)
/home/alan/OpencOR/src/plugins/support/CellMLSupport/src/cellmlsupportplugin.cpp:103: Class 'OpenCOR::CellMLSupport' lacks Q_OBJECT macro
/home/alan/OpencOR/src/plugins/support/CellMLSupport/src/cellmlsupportplugin.cpp:48: Unbalanced opening parenthesis in C++ code (or abuse of the C++ preprocessor)
/home/alan/OpencOR/src/plugins/support/COMBINESupport/src/combinearchive.cpp:297: Excess closing parenthesis in C++ code (or abuse of the C++ preprocessor)
/home/alan/OpencOR/src/plugins/support/COMBINESupport/src/combinearchive.cpp:388: Class 'OpenCOR::COMBINESupport' lacks Q_OBJECT macro
/home/alan/OpencOR/src/plugins/support/COMBINESupport/src/combinearchive.cpp:297: Unbalanced opening parenthesis in C++ code (or abuse of the C++ preprocessor)
/home/alan/OpencOR/src/plugins/support/COMBINESupport/src/combinesupportplugin.cpp:94: Class 'OpenCOR::COMBINESupport' lacks Q_OBJECT macro
/home/alan/OpencOR/src/plugins/support/COMBINESupport/src/combinesupportplugin.cpp:40: Unbalanced opening parenthesis in C++ code (or abuse of the C++ preprocessor)
/home/alan/OpencOR/src/plugins/support/PMRSupport/src/pmrsupportplugin.cpp:43: Unbalanced opening parenthesis in C++ code (or abuse of the C++ preprocessor)
/home/alan/OpencOR/src/plugins/support/PMRSupport/src/pmrsupportpreferenceswidget.cpp:135: Unbalanced opening parenthesis in C++ code (or abuse of the C++ preprocessor)
/home/alan/OpencOR/src/plugins/support/SEDMLSupport/src/sedmlsupportplugin.cpp:102: Class 'OpenCOR::SEDMLSupport' lacks Q_OBJECT macro
/home/alan/OpencOR/src/plugins/support/SEDMLSupport/src/sedmlsupportplugin.cpp:47: Unbalanced opening parenthesis in C++ code (or abuse of the C++ preprocessor)
/home/alan/OpencOR/src/plugins/tools/CellMLTools/src/cellmltoolsplugin.cpp:161: Class 'OpenCOR::CellMLTools' lacks Q_OBJECT macro
/home/alan/OpencOR/src/plugins/tools/CellMLTools/src/cellmltoolsplugin.cpp:54: Unbalanced opening parenthesis in C++ code (or abuse of the C++ preprocessor)
/home/alan/OpencOR/src/plugins/widget/MathMLViewerWidget/src/mathmlviewerwidgetplugin.cpp:37: Unbalanced opening parenthesis in C++ code (or abuse of the C++ preprocessor)
-- Configuring done
-- Generating done
-- Build files have been written to: /home/alan/OpencOR/build
...
```

i.e. a lot of `Q_OBJECT` related issues while we don't get them using Qt 5.12.3...!?!?

**Comment by agarny** [comment]
I don't know whether it's related, but a few weeks ago (on 21 April) we set `CMAKE_AUTOMOC` and `CMAKE_AUTOUIC` to `ON` in our main `CMakeLists.txt` (see [here](https://github.com/opencor/opencor/commit/7037071a0387b3466f28fa4ce1cbd26d9705cd3c#diff-af3b638bc2a3e6c650974192a53c7291R246)).

So... could that be the cause of some/all our problems?...

**Comment by agarny** [comment]
Ok, I have just built and tested a Windows and a Linux version of commit 0fa5695 (i.e. just before the commit that includes `CMAKE_AUTOMOC` and `CMAKE_AUTOUIC` to `ON`), and it's all working fine.

Then, I did the same with commit 7037071 (i.e. the commit that includes `CMAKE_AUTOMOC` and `CMAKE_AUTOUIC` to `ON`), and it's NOT working.

Bottom line: for some unclear reason, to set `CMAKE_AUTOMOC` and `CMAKE_AUTOUIC` to `ON` has broken OpenCOR on Windows and Linux...!?

So, ok, it looks like I am going to revert that aspect of commit 7037071. I guess it was a typical case of "if it ain't broke, don't fix it"... Ok, tomorrow is another day!

**Comment by agarny** [comment]
Closing in favour of issue #2094.

**Comment by agarny** [comment]
Reopening this issue since issue #2094 assumed that the problem is with `CMAKE_AUTO*`, but it's not the case in the end.

All we know is that commit 77cd1f2 is the commit just before we introduced our use of `CMAKE_AUTO*` and that everything works fine it while commit 7037071 is the commit where we introduced `CMAKE_AUTO*` and the point from which things are broken.

Now, the problem is that commit 7037071 introduces a lot of changes (as a result of our use of Clang-Tidy) and, clearly, something in that commit breaks OpenCOR and it's not our use of `CMAKE_AUTO*`.
